karapace/karapacemetrics.py

```diff
@@ -0,0 +1,169 @@
+        sensor.add(MetricName("response-size-avg", "kafka-metrics"), Avg())
+
+        sensor = self.metrics.sensor("master-slave-role")
+        sensor.add(MetricName("master-slave-role", "kafka-metrics"), Value())
+
+        sensor = self.metrics.sensor("request-error-rate")
+        sensor.add(MetricName("request-error-rate", "kafka-metrics"), Rate())
+
+        sensor = self.metrics.sensor("request-rate")
+        sensor.add(MetricName("request-rate", "kafka-metrics"), Rate())
+
+        sensor = self.metrics.sensor("response-rate")
+        sensor.add(MetricName("response-rate", "kafka-metrics"), Rate())
+
+        sensor = self.metrics.sensor("response-byte-rate")
+        sensor.add(MetricName("response-byte-rate", "kafka-metrics"), Rate())
+
+        sensor = self.metrics.sensor("latency")
+        sensor.add(MetricName("latency-max", "kafka-metrics"), Max())
+        sensor.add(MetricName("latency-avg", "kafka-metrics"), Avg())
+
+        self.stats_client = stats_client
+
+        schedule.every(10).seconds.do(self.schedule)
+
+        self.worker_thread.start()
+
+    def connection(self) -> None:
+        if not self.active:
+            return
+        timestamp = int(datetime.utcnow().timestamp() * 1e3)
+        self.metrics.get_sensor("connections-active").record(1.0, timestamp)
+
+    def request(self, size: int) -> None:
+        if not self.active:
+            return
+        timestamp = int(datetime.utcnow().timestamp() * 1e3)
+        self.metrics.get_sensor("request-size").record(size, timestamp)
+        self.metrics.get_sensor("request-rate").record(1, timestamp)
+
+    def response(self, size: int) -> None:
+        if not self.active:
+            return
+        timestamp = int(datetime.utcnow().timestamp() * 1e3)
+        self.metrics.get_sensor("connections-active").record(-1.0, timestamp)
+        self.metrics.get_sensor("response-size").record(size, timestamp)
+        self.metrics.get_sensor("response-byte-rate").record(size, timestamp)
+        self.metrics.get_sensor("response-rate").record(1, timestamp)
+
+    def are_we_master(self, is_master: bool) -> None:
+        if not self.active:
+            return
+        timestamp = int(datetime.utcnow().timestamp() * 1e3)
+        self.metrics.get_sensor("master-slave-role").record(int(is_master), timestamp)
+
+    def latency(self, latency_ms: float) -> None:
+        if not self.active:
+            return
+        timestamp = int(datetime.utcnow().timestamp() * 1e3)
+        self.metrics.get_sensor("latency").record(latency_ms, timestamp)
+
+    def error(self) -> None:
+        if not self.active:
+            return
+        timestamp = int(datetime.utcnow().timestamp() * 1e3)
+        self.metrics.get_sensor("request-error-rate").record(1, timestamp)
+
+    def report(self) -> None:
+        if not self.active:
+            return
+        if isinstance(self.stats_client, StatsClient):
```

Comment:
issue: I think this should be a `RuntimeError` as it shouldn't be possible for `self.stats_client` not be a `StatsClient` when `self.active` is `True`.

Reply:
fixed